### PR TITLE
Add a guide about adding a SGB border

### PR DESCRIPTION
**Pull request by zlago**
closes gbdev/awesome-gbdev#246

### CODE REVIEW

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:3]:
```suggestion

```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:1]:
```suggestion
# Adding a SGB border to your non-SGB game
```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:5]:
```suggestion
Table of contents:
```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:18]:
```suggestion
Before we can do anything, we must first specify in the header that this game is an SGB game.
```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:20]:
```suggestion
To enable SGB features:

```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:22]:
```suggestion
- Old licensee code (`$014b`) must be set to `$33`
```

File: website/guides/border.md
Comment (avivace) [inline website/guides/border.md:24]:
```suggestion
## Packets
```

**Review by avivace — CHANGES_REQUESTED** [review]

File: website/guides/border.md
Comment (pinobatch) [reply website/guides/border.md:1]: "an ess-jee-bee" or "a Super Game Boy"?

Comment (zlago) [reply website/guides/border.md:1]: i intentionally used an initialism there - it should be "an SGB".

File: website/guides/sgb_border.md
Comment (avivace) [inline website/guides/sgb_border.md:169]: we should just redirect to discord here (link to the chat channels on the gbdev.io website)

**Review by avivace — CHANGES_REQUESTED** [review]

**Review by avivace — CHANGES_REQUESTED** [review]

File: website/guides/sgb_border.md
Comment (zlago) [reply website/guides/sgb_border.md]: whoops forgot about this suggestion ><

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:28]: It has a comma splice after the change.
```diff
-first), if
+first). If
```

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:18]: Is it that [bit banging](https://en.wikipedia.org/wiki/Bit_banging) is not well known as a term of art for toggling GPIOs, or some other reason?

File: website/guides/sgb_border.md
Comment (zlago) [reply website/guides/sgb_border.md:28]: this is the type of thing that makes me go
```
git ci --amend --no-e
git fpush
```
then
```
git add -u
git ci --amend --no-e
git fpush
```
after i realise i forgot to stage ><

File: website/guides/sgb_border.md
Comment (zlago) [reply website/guides/sgb_border.md:18]: ..."Communicating with the SGB is done by sending a bitstream through the `P1` register"?

Comment (avivace) [reply website/guides/sgb_border.md:18]: > Is it that [bit banging](https://en.wikipedia.org/wiki/Bit_banging) is not well known as a term of art for toggling GPIOs, or some other reason?

"Bit banging" is jargon, which should be avoided when not necessary

Comment (zlago) [reply website/guides/sgb_border.md:18]: does LSB/MSB not count as jargon? would you approve of `sent bit-by-bit ("bitbanged")` or `[bitbanged](wikipedia link)`?

File: website/guides/sgb_border.md
Comment (avivace) [reply website/guides/sgb_border.md:10]: Good idea, other guides have a similar introduction (not necessarily in bullet points)

File: website/guides/sgb_border.md
Comment (avivace) [inline website/guides/sgb_border.md:51]: This needs to be reworded. Maybe mention what do we mean by "TRN" here?

File: website/guides/sgb_border.md
Comment (avivace) [inline website/guides/sgb_border.md:47]: Minor: should additional references/links be put here (e.g. a use case/example of sending more than 16 bytes?) or is it irrelevant/superficial?

File: website/guides/sgb_border.md
Comment (avivace) [inline website/guides/sgb_border.md:60]: Should we include the snippet here as we did with the other one?

**Review by avivace — APPROVED** [review]
I finished reviewing, I left some minor comments but I believe it's in a merge-able state already ;)

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:20]:
```suggestion
Before we can do anything else, we must first specify in the header that this game is aware of SGB features.
Otherwise, the SGB BIOS will ignore any packets we send, and we won't even be able to detect when the program is running on SGB.
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:5]:
```suggestion
This document was written to aid developers of DMG compatible homebrew with adding Super Game Boy borders.

We will see how to:
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:51]:
```suggestion
Bulk transfer (TRN) packets tell the SGB to copy the contents of the screen to buffers in Super NES work RAM. The `CHR_TRN` and `PCT_TRN` packets are used to send data for SGB borders.
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:57]:
```suggestion
3. the tilemap consists of `$00`, `$01`..`$13`, 12 bytes padding (offscreen), `$14`..`$27`, padding, repeat until `$ff` (inclusive)
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:76]:
```suggestion
- attempt to advance the read player (set `P1.5` to 0 then 1)
- set `P1.4`-`P1.5` to `%11` (`%xx11xxxx`)
- read `P1.0`-`P1.3`, and check if it either
	* has changed
	* isn't `%1111`
- advance and reread a few times and branch somewhere if the test keeps failing
```
"reset" is less clear than "set to 0", and "repeat a few times" could mean repeat the `MLT_REQ` too

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:63]:
```suggestion
- **You must load the data into VRAM and enable the screen before sending the TRN packet**
- **You must wait ~8 frames after each TRN** instead of just 4
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:144]:
```suggestion
- `--color-zero` should be the color that your image for transparency, in my case it was blue.
	* If your image has an alpha channel, it can be set can also be set to `00000000` to use the actual transparent color; however, this may cause some issues.
- `-v` is optional, for showing details of the conversion process
```

File: website/guides/sgb_border.md
Comment (pinobatch) [inline website/guides/sgb_border.md:174]:
```suggestion
2. You can use palettes 0-3 and 7, if you really know what you're doing (animated borders yay). You will probably have to edit the border in YY-CHR, as there aren't really any other tools for that.
```
Unlike some other languages, English doesn't capitalize "you" as if it were a proper noun.

**Review by pinobatch — CHANGES_REQUESTED** [review]
Mostly spelling and grammar fixes

File: website/guides/sgb_border.md
Comment (ISSOtm) [reply website/guides/sgb_border.md]: I wanted to mention that what controls the "enabling of SGB features" is the ROM's header, and not e.g. code. I figured that it could help with troubleshooting?

File: website/guides/border.md
Comment (nitro2k01) [inline website/guides/border.md]: This is perfectly fine code for using in a game, but I don't think it's good tutorial code because it drags in a bunch of implicit assumptions that someone who is learning may not be familiar with. It's using `halt` to wait for interrupts but this relies on the specific setup listed here:

- Interrupts must be enabled during this section, whereas interrupts are often disabled during setup code.
- LCDC must be on during this code in order to generate the interrupts, whereas LCDC is often disabled during initialization to make VRAM accessible.
- It relies in the specific setup of `STAT` which is given here as "magic values".
- It relies on the STAT interrupt being a single `reti` or similar.

This (imo) makes this code fragile for use by a beginner unless they understand how all the pieces fit together. (By fragile I mean that a change somewhere else can easily cause this code to stop working.) It would be easy for me, who has been doing this for 20 years at this point, to say that this is good code because I can understand, see that it works, or otherwise know exactly what to do if it doesn't work. But I'm not the target audience of this code.

I would much prefer the more "traditional" approach of a busy wait loop for waiting frames, which, while not as fancy, solves all of the issues listed above.

**Review by nitro2k01 — CHANGES_REQUESTED** [review]
I'd ask that the example code for testing for SGB use a busy wait loop instead of interrupts because the current example code makes assumptions which may be confusing for a beginner.

File: website/guides/border.md
Comment (zlago) [reply website/guides/border.md]: - doesnt the doc specify that LCD must be enabled during TRNs?
- arent waitloops that a beginner would use ineffective if LCD is off?
- are you not going to comment on the doc assuming you have means of VRAM-safe VRAM access?

Comment (nitro2k01) [reply website/guides/border.md]: > doesnt the doc specify that LCD must be enabled during TRNs?

Sure, but this is something they must keep in mind if they might later rearrange the code to put the detection in a section with LCDC disabled. And that doesn't address any of the other issues. For example, I'm assuming you chose to use STAT interrupt instead of the more obvious VBlank interrupt in order to avoid extra complexity in the VBlank interrupt handler. But now the STAT interrupt is "used up" instead. Using a waitloop makes this code more self contained without forcing assumptions on other parts of the code.

> arent waitloops that a beginner would use ineffective if LCD is off?

I'm talking about a waitloop with a counter, not a `LY` waitloop. Something like Nintendo's suggestion:
```sm83asm
    ld de,$1b58
:   nop
    nop
    nop
    dec de
    ld a,d
    or e
    jr nz,:-
    ret
```

That particular code has some awkward `nop`s in there, seemingly so that the loop takes exactly 10 M cycles and spins 7000 times for a total of 70000 M cycles for easy math. But that's the general idea.

> are you not going to comment on the doc assuming you have means of VRAM-safe VRAM access?

Safe VRAM access in itself is one of the most central concepts in GB programming, so someone who is new will be presumed to learn it sooner rather than later.

Comment (zlago) [reply website/guides/border.md]: actually, i already happened to have STAT used up for doing nothing in snek-gbc (which is where all the snippets are from), in every other project i use vblank since there the handler only does what its told to (snek-gbc was programmed with the assumption that i would never have to build upon it, dont ask me why i felt the *need* to include a dmg-only startup animation)

File: website/guides/sgb_border.md
Comment (avivace) [reply website/guides/sgb_border.md]: Uh, okay, I see your point.. We should then try to simplify it a bit (e.g. to help restrict the cases in which really this "troubleshooting" may happen and the user should be aware of the fact it may involve SGB side, if I understood you correctly)

File: website/guides/border.md
Comment (pinobatch) [reply website/guides/border.md]: So to summarize, are these the requested changes?
- Do the delay with `di` and a busy wait loop instead of vblank/STAT interrupt
- Emphasize that the display must be on, or the SGB will never receive the data

Comment (avivace) [reply website/guides/border.md]: yep, @pinobatch , sounds like it ;)

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:71]: This sounds confusing. What does it mean?

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:114]: `sm83asm` is not recognised by any highlighter out there; I don't know which one is used on gbdev.io, but `rgbasm` is supported by highlight.js [via a plugin](https://github.com/gbdev/highlightjs-rgbasm), and alternatives like `avrasm`, `z80asm`, `x86asm`, or `nasm` degrade well enough.

Also, "block" comments (`/* */`) are very uncommon in assembly syntaxes; single-line comments (`;`) should be used instead, especially if using one of the alternatives listed above.

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:116]: The label should probably be called `.notSgb` then.

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:168]:
```suggestion
[^1]: You can send a `CHR_TRN` [up to \~60 frames](https://github.com/pinobatch/little-things-gb/blob/master/sgbears/docs/long_story.txt) after the `PCT_TRN` for it to apply to the current border, but not all emulators will emulate this. It's fine to just pretend `CHR_TRN`s must go before `PCT_TRN`.
```

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:174]: What does "if you really know what you're doing" imply?

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:176]:
```suggestion
3. When the SNES lags, scanline 225 of the SGB border will be visible! You can set the topmost row of the 29th row of tiles to black to hide this.
```

File: website/guides/sgb_border.md
Comment (ISSOtm) [inline website/guides/sgb_border.md:178]:
```suggestion
4. If this doesn't work for you, you can ask for help on the [gbdev](https://gbdev.io/chat.html) channels.
```

**Review by ISSOtm — CHANGES_REQUESTED** [review]
Turns out these didn't submit when I first made them. Nice.

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:177]: There may be advanced techniques to draw parts of the border with the game palette (0) instead of the palettes reserved for the border (4-6). Currently no converter supports doing this. Would it be worthwhile to make a test ROM to confirm that it's possible?

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:171]: I might change the name of `long_story.txt` in a future revision of this repository. Sometimes I leave things as `.txt` while writing them until I finish enough of the first draft to be able to tell whether or not I'll be using features that make the Markdown source code less readable in a text editor. Should this tutorial link to `long_story.txt` in a particular revision, or should it wait until I've made a decision about renaming it?

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:74]: My guess:
1. wait for the SGB to finish initializing
2. set multiplayer mode and wait for the SGB to recognize the request to go to multiplayer mode
3. test whether multiplayer mode was activated by reading the controller a couple times and seeing whether the input register reflects a player other than player 1
4. optionally turn off multiplayer mode

File: website/guides/sgb_border.md
Comment (avivace) [reply website/guides/sgb_border.md:117]: Agree on both points here

File: website/guides/sgb_border.md
Comment (avivace) [reply website/guides/sgb_border.md:171]: The link right now it's broken for me. We should link to it in a particular git commit so the link won't break and you can PR the new link if/when it changes.

File: website/guides/sgb_border.md
Comment (zlago) [reply website/guides/sgb_border.md:177]: since its not that practical to use the game palette, it should be a low priority

File: website/guides/sgb_border.md
Comment (pinobatch) [reply website/guides/sgb_border.md:171]: Link to a specific commit <https://github.com/pinobatch/little-things-gb/blob/b11b554d73c48a0f54fee0df31e59eb83806fcb4/sgbears/docs/long_story.md>

File: website/guides/sgb_border.md
Comment (ISSOtm) [reply website/guides/sgb_border.md:177]: Then I think this shouldn't be mentioned in the tutorial, beyond something like "palette 0 is used by the displayed GB screen, palettes 1-3 and 7 are used by the SGB BIOS and will be overwritten when opening various menus".

Mentioning that they *can* be used, but not what the limitations are, is likely to be more confusing.

File: website/guides/sgb_border.md
Comment (ISSOtm) [reply website/guides/sgb_border.md:171]: Thanks Pino! Updated my original suggestion accordingly.

**Comment by pinobatch** [comment]
It looks like [this comment](https://discord.com/channels/303217943234215948/889588664328409109/1161791405589528707) describes the last things that some contributor to this pull request needs to do before a maintainer can merge this pull request. In order to do them, I'd need to start yet another review creating yet more suggestions. However, in the "Files changed" tab, there are still a bunch of suggestions, some mine, that are marked "avivace marked this conversation as resolved" yet there was no action on the suggestion. I'm not sure what happened, nor whether I'm supposed to be making more suggestions on top of my previous suggestions that are "marked [...] as resolved" without either acceptance or any discussion of rejection.

File: website/guides/sgb_border.md
Comment (avivace) [reply website/guides/sgb_border.md:74]: @ISSOtm

**Comment by avivace** [comment]
I'll go ahead and merge what we have now as the threads got confusing. Let's continue in another PR.

**Pull request merged**
